File: src/read.ts
import { createDiagnostics } from "./diagnostics.ts";
import { detectHarnesses, type Detection } from "./detect.ts";
import { native } from "./native.ts";
import {
  createPathResolver,
  effectiveCwd,
  effectiveEnv,
  effectiveHome,
  type PathResolver,
  type PathResolverOptions,
} from "./paths.ts";
import type { Diagnostic, Harness, UsageEvent } from "./types.ts";

export type ReadUsageOptions = PathResolverOptions & {
  harnesses?: Harness[];
  // ISO date or Date; events strictly before are dropped
  // date-only strings (YYYY-MM-DD) mean midnight of that day in tz
  since?: Date | string;
  // ISO date or Date; events at/after are dropped
  // date-only strings (YYYY-MM-DD) mean the exclusive end of that day in tz
  until?: Date | string;
  // IANA timezone for date-only since/until bounds; defaults to the system timezone
  tz?: string;
  subagents?: "include" | "exclude";
};

export type ReadUsageResult = {
  events: UsageEvent[];
  diagnostics: Diagnostic[];
  // per-harness source files (or file:line) skipped as unreadable/malformed
  skipped: Partial<Record<Harness, string[]>>;
};

const DATE_ONLY = /^(\d{4})-(\d{2})-(\d{2})$/;
const DAY_MS = 86_400_000;

function tzOffsetMs(tz: string | undefined, at: number): number {
  const options: Intl.DateTimeFormatOptions = {
    era: "short",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hourCycle: "h23",
  };
  // an unknown zone degrades to UTC rather than aborting the read, matching the
  // native core's date-only bound resolution
  let parts: Intl.DateTimeFormatPart[];
  try {
    parts = new Intl.DateTimeFormat("en-CA", {
      ...(tz !== undefined ? { timeZone: tz } : {}),
      ...options,
    }).formatToParts(at);
  } catch {
    parts = new Intl.DateTimeFormat("en-CA", { timeZone: "UTC", ...options }).formatToParts(at);
  }
  const get = (type: Intl.DateTimeFormatPartTypes): number => {
    const part = parts.find((p) => p.type === type);
    if (part === undefined) throw new RangeError(`missing ${type} in formatted date`);
    return Number(part.value);
  };
  const era = parts.find((p) => p.type === "era")?.value ?? "";
  const year = era.startsWith("B") ? 1 - get("year") : get("year");
  const asUtc = new Date(0);
  asUtc.setUTCFullYear(year, get("month") - 1, get("day"));
  asUtc.setUTCHours(get("hour"), get("minute"), get("second"));
  return asUtc.getTime() - at;
}

// resolve a wall-clock midnight in tz to an epoch by sampling nearby offsets;
// an ambiguous midnight (DST overlap) takes its first occurrence, and a
// midnight inside a DST gap takes the first valid instant of the day
function zonedMidnight(
  value: string,
  match: RegExpExecArray,
  tz: string | undefined,
  dayOffset: number,
): number {
  const year = Number(match[1]);
  const month = Number(match[2]);
  const day = Number(match[3]);
  const roundtrip = new Date(0);
  roundtrip.setUTCFullYear(year, month - 1, day);
  const base = roundtrip.getTime();
  if (
    roundtrip.getUTCFullYear() !== year ||
    roundtrip.getUTCMonth() !== month - 1 ||
    roundtrip.getUTCDate() !== day
  ) {
    throw new RangeError(`invalid date: ${value}`);
  }
  const guess = base + dayOffset * DAY_MS;
  const offsets = new Set([
    tzOffsetMs(tz, guess - DAY_MS),
    tzOffsetMs(tz, guess),
    tzOffsetMs(tz, guess + DAY_MS),
  ]);
  const candidates = [...offsets].map((offset) => guess - offset);
  const valid = candidates.filter((candidate) => candidate + tzOffsetMs(tz, candidate) === guess);
  if (valid.length > 0) return Math.min(...valid);
  return Math.max(...candidates);
}

function toTime(
  value: Date | string | undefined,
  tz: string | undefined,
  exclusiveEndOfDay: boolean,
): number | null {
  if (value === undefined) return null;
  if (typeof value === "string") {
    const match = DATE_ONLY.exec(value);
    if (match !== null) return zonedMidnight(value, match, tz, exclusiveEndOfDay ? 1 : 0);
  }
  const time = value instanceof Date ? value.getTime() : Date.parse(value);
  if (!Number.isFinite(time)) throw new RangeError(`invalid date: ${String(value)}`);
  return time;
}

export function detect(options: PathResolverOptions = {}): Detection {
  return detectHarnesses(createPathResolver(options));
}

// The addon cannot see the parent Node process env, so the facade resolves the
// effective {home, env} on the TS side and passes them down as concrete data.
async function gatherHarness(
  harness: Harness,
  options: ReadUsageOptions,
): Promise<{ events: UsageEvent[]; diagnostics: Diagnostic[]; skipped: string[] }> {
  // empty-string values are dropped so the addon's "empty means unset" rule
  // agrees with the resolver
  const env: Record<string, string> = {};
  for (const [name, value] of Object.entries(effectiveEnv(options))) {
    if (value !== undefined && value !== "") env[name] = value;
  }
  const optionsJson = JSON.stringify({
    home: effectiveHome(options),
    env,
    cwd: effectiveCwd(options),
  });
  const raw = await native().readHarness(harness, optionsJson);
  return JSON.parse(raw) as {
    events: UsageEvent[];
    diagnostics: Diagnostic[];
    skipped: string[];
  };
}

export async function readUsage(options: ReadUsageOptions = {}): Promise<ReadUsageResult> {
  const resolver: PathResolver = createPathResolver(options);
  const since = toTime(options.since, options.tz, false);
  const until = toTime(options.until, options.tz, true);
  const excludeSubagents = options.subagents === "exclude";
  const harnesses = options.harnesses ?? detectHarnesses(resolver).supported;

  const diagnostics = createDiagnostics();
  const events: UsageEvent[] = [];
  const skipped: Partial<Record<Harness, string[]>> = {};

  for (const harness of harnesses) {
    const result = await gatherHarness(harness, options);
    // reader warnings are surfaced before this harness's per-event filtering,
    // matching the original in-process order (warn sink fires during the read)
    for (const diagnostic of result.diagnostics)
      diagnostics.add(diagnostic.severity, diagnostic.message, diagnostic);
    if (result.skipped.length > 0) skipped[harness] = result.skipped;
    for (const event of result.events) {
      // one timestamp policy regardless of filtering: an unparsable timestamp
      // drops the event with a diagnostic, never a silent pass-through
      const at = Date.parse(event.timestamp);
      if (Number.isNaN(at)) {
        diagnostics.warn(
          `dropping event ${event.messageId} with unusable timestamp ${JSON.stringify(event.timestamp)}`,
          { harness },
        );
        continue;
      }
      if (since !== null && at < since) continue;
      if (until !== null && at >= until) continue;
      if (excludeSubagents && event.subagent) continue;
      events.push(event);
    }
  }

  return { events, diagnostics: diagnostics.list(), skipped };
}

File: crates/skopli-wire/src/read.rs
//! The multi-reader `read_usage` envelope and harness detection shared by the
//! direct-core bindings that own the whole read+filter loop (`skopli-py`,
//! `skopli-ruby`). `skopli-node` instead exposes a per-harness raw
//! gather (`read_harness`) and runs the filter loop in its TS facade, so it does
//! NOT use this module.
//!
//! Seam split: `{home, env}` + since/until/tz travel as
//! data in the options JSON; a date-only bound resolves to midnight of that day
//! in the requested `tz` (via the bundled tzdb, DST-aware), so date filtering
//! and day rollups agree on where a calendar day begins. An omitted `tz`
//! resolves the system timezone (matching the TS default); `"UTC"` and any
//! unknown zone resolve to UTC midnight. Pricing
//! source loading / fetch / cache live in the language facade, which hands
//! pre-fetched catalog JSON to `create_pricing`; nothing here touches the
//! network.

use serde_json::{Value, json};
use skopli_core::readers::reader::ReaderContext;
use skopli_core::readers::registered_readers;
use skopli_core::types::UsageEvent;

/// Options accepted by `read_usage`, parsed from the options JSON.
pub struct ReadOptions {
    pub ctx: ReaderContext,
    pub harnesses: Option<Vec<String>>,
    pub since_ms: Option<i64>,
    pub until_ms: Option<i64>,
    pub exclude_subagents: bool,
}

/// Parse `read_usage` options from the JSON value. Bad `since`/`until` yield an
/// `Err(message)` that the caller turns into an `InvalidArgumentError`.
pub fn parse_read_options(opts: &Value) -> Result<ReadOptions, String> {
    let ctx = crate::parse::context_from_options(opts);

    let harnesses = opts.get("harnesses").and_then(Value::as_array).map(|arr| {
        arr.iter()
            .filter_map(|v| v.as_str().map(str::to_owned))
            .collect::<Vec<_>>()
    });

    let tz = opts.get("tz").and_then(Value::as_str);
    let since_ms = parse_bound(opts.get("since"), tz, false)?;
    let until_ms = parse_bound(opts.get("until"), tz, true)?;
    let exclude_subagents = opts.get("subagents").and_then(Value::as_str) == Some("exclude");

    Ok(ReadOptions {
        ctx,
        harnesses,
        since_ms,
        until_ms,
        exclude_subagents,
    })
}

/// Resolve a since/until bound to epoch ms. Mirrors `toTime` in read.ts: a
/// date-only `YYYY-MM-DD` string is midnight of that day in `tz` (DST-aware, via
/// the bundled tzdb; an omitted `tz` uses the system zone, `"UTC"`/unknown zone
/// -> UTC midnight), matching the zone the rollup day bucketing uses. Any other
/// string is parsed as ISO.
/// `exclusive_end` shifts a date-only `until` to midnight of the following day.
/// An unparseable value is an error.
fn parse_bound(
    value: Option<&Value>,
    tz: Option<&str>,
    exclusive_end: bool,
) -> Result<Option<i64>, String> {
    let Some(value) = value else {
        return Ok(None);
    };
    if value.is_null() {
        return Ok(None);
    }
    let s = value
        .as_str()
        .ok_or_else(|| "since/until must be a date string".to_owned())?;
    let trimmed = s.trim();
    if trimmed.is_empty() {
        return Ok(None);
    }
    if let Some((y, m, d)) = parse_date_only(trimmed) {
        let day_offset = if exclusive_end { 1 } else { 0 };
        return zoned_midnight(y, m, d, tz, day_offset)
            .ok_or_else(|| format!("invalid date: {s}"))
            .map(Some);
    }
    match skopli_core::readers::shared::date_parse_ms(trimmed) {
        Some(ms) => Ok(Some(ms)),
        None => Err(format!("invalid date: {s}")),
    }
}

/// Parse a bare `YYYY-MM-DD` (no time component) into `(year, month, day)`, or
/// `None` if it is not in that exact form.
fn parse_date_only(s: &str) -> Option<(i16, i8, i8)> {
    let bytes = s.as_bytes();
    if bytes.len() != 10 || bytes[4] != b'-' || bytes[7] != b'-' {
        return None;
    }
    if !bytes
        .iter()
        .enumerate()
        .all(|(i, &b)| i == 4 || i == 7 || b.is_ascii_digit())
    {
        return None;
    }
    let year: i16 = s[0..4].parse().ok()?;
    let month: i8 = s[5..7].parse().ok()?;
    let day: i8 = s[8..10].parse().ok()?;
    Some((year, month, day))
}

/// Midnight of `year-month-day` (plus `day_offset` days) in `tz`, as epoch ms.
/// An omitted `tz` resolves the system IANA zone (matching the TS default of the
/// process system timezone; jiff falls back to UTC if the system zone is
/// undetectable). `"UTC"` and any unknown zone use UTC. An ambiguous wall-clock
/// midnight (a DST fold) takes its first occurrence and a midnight inside a DST
/// gap takes the first valid instant of the day, matching read.ts (jiff's
/// `Compatible` disambiguation). Returns `None` for a calendar-invalid date.
fn zoned_midnight(year: i16, month: i8, day: i8, tz: Option<&str>, day_offset: i64) -> Option<i64> {
    let zone = match tz {
        None => jiff::tz::TimeZone::system(),
        Some("UTC") => jiff::tz::TimeZone::UTC,
        Some(name) => jiff::tz::TimeZone::get(name).unwrap_or(jiff::tz::TimeZone::UTC),
    };
    let date = jiff::civil::Date::new(year, month, day).ok()?;
    let date = date.checked_add(jiff::Span::new().days(day_offset)).ok()?;
    let ts = zone.to_timestamp(date.at(0, 0, 0, 0)).ok()?;
    Some(ts.as_millisecond())
}

/// The `read_usage` result envelope: `{events, diagnostics, skipped}`. Events
/// are the union across the selected harnesses, in per-harness read order.
pub fn read_usage(options: &ReadOptions) -> Value {
    let selected: Option<std::collections::HashSet<&str>> = options
        .harnesses
        .as_ref()
        .map(|h| h.iter().map(String::as_str).collect());

    let mut events: Vec<UsageEvent> = Vec::new();
    let mut diagnostics: Vec<Value> = Vec::new();
    let mut skipped: serde_json::Map<String, Value> = serde_json::Map::new();

    for reader in registered_readers() {
        let harness = reader.harness_id();
        if let Some(sel) = &selected
            && !sel.contains(harness)
        {
            continue;
        }
        let result = reader.read(&options.ctx);

        for warning in &result.warnings {
            diagnostics.push(json!({
                "severity": "warning",
                "message": warning.message.trim_end_matches('\n'),
                "harness": harness,
            }));
        }
        if !result.skipped.is_empty() {
            skipped.insert(harness.to_owned(), json!(result.skipped));
        }

        for event in result.events {
            match skopli_core::readers::shared::date_parse_ms(&event.timestamp) {
                None => {
                    diagnostics.push(json!({
                        "severity": "warning",
                        "message": format!(
                            "dropping event {} with unusable timestamp {}",
                            event.message_id,
                            serde_json::to_string(&event.timestamp).unwrap_or_default()
                        ),
                        "harness": harness,
                    }));
                    continue;
                }
                Some(at) => {
                    if let Some(since) = options.since_ms
                        && at < since
                    {
                        continue;
                    }
                    if let Some(until) = options.until_ms
                        && at >= until
                    {
                        continue;
                    }
                    if options.exclude_subagents && event.subagent {
                        continue;
                    }
                    events.push(event);
                }
            }
        }
    }

    let events_json: Vec<Value> = events
        .iter()
        .map(|e| serde_json::to_value(e).expect("serialize event"))
        .collect();

    json!({
        "events": events_json,
        "diagnostics": diagnostics,
        "skipped": Value::Object(skipped),
    })
}

/// Detect which registered harnesses have readable data reachable from the
/// context. A harness is `supported` when its reader produced at least one
/// event; `unsupported` has no core representation and is always empty.
pub fn detect(opts: &Value) -> Value {
    let ctx = crate::parse::context_from_options(opts);
    let mut supported: Vec<String> = Vec::new();
    for reader in registered_readers() {
        let result = reader.read(&ctx);
        if !result.events.is_empty() {
            supported.push(reader.harness_id().to_owned());
        }
    }
    json!({
        "supported": supported,
        "unsupported": Value::Array(Vec::new()),
    })
}
